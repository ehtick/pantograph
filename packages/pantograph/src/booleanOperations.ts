import {
  cutFiguresLists,
  fuseFiguresLists,
  intersectFiguresLists,
} from "./algorithms/boolean/figureBooleans";
import { Diagram } from "./models/Diagram";
import { Figure } from "./models/Figure";
import { Loop } from "./models/Loop";
import { listOfFigures } from "./utils/listOfFigures";

export function fuse(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    fuseFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}

export function fuseAll(shapes: (Diagram | Figure | Loop)[]): Diagram {
  return shapes.reduce(
    (acc: Diagram, shape: Diagram | Figure | Loop) => fuse(acc, shape),
    new Diagram()
  );
}

export function cut(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    cutFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}

export function intersect(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    intersectFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}
