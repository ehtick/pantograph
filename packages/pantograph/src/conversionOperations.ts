import { Segment } from "./models/segments/Segment";
import { Arc } from "./models/segments/Arc";
import { Line } from "./models/segments/Line";
import { QuadraticBezier } from "./models/segments/QuadraticBezier";
import { CubicBezier } from "./models/segments/CubicBezier";
import { EllipseArc } from "./models/segments/EllipseArc";
import { approximateBezierCurveAsArcs } from "./algorithms/conversions/bezierToArcs";
import { approximateEllipticalArcAsCubicBeziers } from "./algorithms/conversions/ellipseToBezier";
import { genericConversion } from "./algorithms/conversions/helpers";
import { Diagram, Figure, Loop, Strand } from "./main";

function mapAsArcsAndLines(s: Segment, tolerance = 1e-3): Segment[] {
  if (s instanceof Arc || s instanceof Line) {
    return [s];
  }
  if (s instanceof QuadraticBezier || s instanceof CubicBezier) {
    return approximateBezierCurveAsArcs(s, tolerance);
  }
  if (s instanceof EllipseArc) {
    return approximateEllipticalArcAsCubicBeziers(s).flatMap((s) =>
      approximateBezierCurveAsArcs(s, tolerance),
    );
  }

  throw new Error(`Unknown segment type: ${s}`);
}

function convertToArcsAndLines(shape: Diagram): Diagram;
function convertToArcsAndLines(shape: Figure): Figure;
function convertToArcsAndLines(shape: Loop): Loop;
function convertToArcsAndLines(shape: Strand): Strand;
function convertToArcsAndLines(shape: Segment): Segment[];
function convertToArcsAndLines(
  shape: Diagram | Figure | Loop | Strand | Segment,
  tolerance = 1e-3,
) {
  const mapper = (segment: Segment) => mapAsArcsAndLines(segment, tolerance);
  return genericConversion(shape, mapper);
}

export { convertToArcsAndLines };
