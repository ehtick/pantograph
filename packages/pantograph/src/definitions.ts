export type Vector = [number, number];

export function isVector(vector: unknown): vector is Vector {
  return Array.isArray(vector) && vector.length === 2;
}
