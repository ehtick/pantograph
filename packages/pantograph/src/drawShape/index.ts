export { drawCircle } from "./drawCircle.js";
export { drawRect } from "./drawRect.js";
export { drawEllipse } from "./drawEllipse.js";
export { drawSVGPath } from "./drawSVGPath.js";
