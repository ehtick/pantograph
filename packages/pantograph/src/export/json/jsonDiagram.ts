import { Diagram } from "../../models/Diagram.js";
import { jsonFigure } from "./jsonFigure.js";

export function jsonDiagram(diagram: Diagram) {
  return {
    type: "DIAGRAM",
    figures: diagram.figures.map(jsonFigure),
  };
}
