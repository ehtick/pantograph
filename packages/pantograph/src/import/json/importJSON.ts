import { Line } from "../../models/segments/Line.js";
import { Arc } from "../../models/segments/Arc.js";
import { Loop } from "../../models/Loop.js";
import { Figure } from "../../models/Figure.js";
import { Diagram } from "../../models/Diagram.js";
import { EllipseArc } from "../../models/segments/EllipseArc.js";
import {CubicBezier} from "../../models/segments/CubicBezier.js";

const importSegment = (json: any) => {
  if (json.type === "LINE") {
    return new Line(json.firstPoint, json.lastPoint);
  }
  if (json.type === "ARC") {
    return new Arc(
      json.firstPoint,
      json.lastPoint,
      json.center,
      json.clockwise
    );
  }
  if (json.type === "ELLIPSE_ARC") {
    return new EllipseArc(
      json.firstPoint,
      json.lastPoint,
      json.center,
      json.majorRadius,
      json.minorRadius,
      json.tiltAngle,
      json.clockwise,
      { angleUnits: "rad" }
    );
  }
  if (json.type === "CUBIC_BEZIER") {
    return new CubicBezier(
      json.firstPoint,
      json.lastPoint,
      json.firstControlPoint,
      json.lastControlPoint
    );
  }
  throw new Error("Unknown segment type");
};

const importLoop = (json: any) => {
  const segments = json.segments.map(importSegment);
  return new Loop(segments);
};

const importFigure = (json: any) => {
  const contour = importLoop(json.contour);
  const holes = json.holes.map(importLoop);
  return new Figure(contour, holes);
};

const importDiagram = (json: any) => {
  const figures = json.figures.map(importFigure);
  return new Diagram(figures);
};

export function importJSON(json: any) {
  if (json.type === "DIAGRAM") {
    return importDiagram(json);
  }
  if (json.type === "FIGURE") {
    return importFigure(json);
  }
  if (json.type === "LOOP") {
    return importLoop(json);
  }
  if (
    json.type === "LINE" ||
    json.type === "ARC" ||
    json.type === "ELLIPSE_ARC" ||
    json.type === "CUBIC_BEZIER"
  ) {
    return importSegment(json);
  }
  throw new Error("Unknown shape type");
}
