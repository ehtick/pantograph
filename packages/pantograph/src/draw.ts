import { Vector } from "./definitions";
import { Diagram } from "./models/Diagram";
import { Figure } from "./models/Figure";
import { Loop } from "./models/Loop";
import { Line } from "./models/segments/Line";
import { Segment } from "./models/segments/Segment";
import { TransformationMatrix } from "./models/TransformationMatrix";
import {
  polarToCartesian,
  DEG2RAD,
  subtract,
  sameVector,
} from "./vectorOperations";

function closeSegments(segments: Segment[]) {
  if (!segments.length) throw new Error("No segments to close");
  const firstSegment = segments[0];
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const lastSegment = segments.at(-1)!;

  if (sameVector(firstSegment.firstPoint, lastSegment.lastPoint))
    return segments;
  return [
    ...segments,
    new Line(lastSegment.lastPoint, firstSegment.firstPoint),
  ];
}

function loopySegmentsToDiagram(segments: Segment[]) {
  // Here we will need to do our best to fix cases where the drawing is
  // broken in some way (i.e. self-intersecting loops)

  return new Diagram([new Figure(new Loop(segments))]);
}

export class DrawingPen {
  protected pointer: Vector;
  protected firstPoint: Vector;
  protected pendingSegments: Segment[];

  constructor(origin: Vector = [0, 0]) {
    this.pointer = origin;
    this.firstPoint = origin;

    this.pendingSegments = [];
  }

  movePointerTo(point: Vector): this {
    if (this.pendingSegments.length)
      throw new Error(
        "You can only move the pointer if there is no segment defined"
      );

    this.pointer = point;
    this.firstPoint = point;
    return this;
  }

  protected saveSegment(segment: Segment) {
    this.pendingSegments.push(segment);
    return this;
  }

  lineTo(point: Vector): this {
    const segment = new Line(this.pointer, point);
    this.pointer = point;
    return this.saveSegment(segment);
  }

  line(xDist: number, yDist: number): this {
    return this.lineTo([this.pointer[0] + xDist, this.pointer[1] + yDist]);
  }

  vLine(distance: number): this {
    return this.line(0, distance);
  }

  hLine(distance: number): this {
    return this.line(distance, 0);
  }

  vLineTo(yPos: number): this {
    return this.lineTo([this.pointer[0], yPos]);
  }

  hLineTo(xPos: number): this {
    return this.lineTo([xPos, this.pointer[1]]);
  }

  polarLineTo([r, theta]: Vector): this {
    const angleInRads = theta * DEG2RAD;
    const point = polarToCartesian(r, angleInRads);
    return this.lineTo(point);
  }

  polarLine(distance: number, angle: number): this {
    const angleInRads = angle * DEG2RAD;
    const [x, y] = polarToCartesian(distance, angleInRads);
    return this.line(x, y);
  }

  tangentLine(distance: number): this {
    const previousCurve = this.pendingSegments.at(-1);

    if (!previousCurve)
      throw new Error("You need a previous segment to sketch a tangent line");

    const [xDir, yDir] = previousCurve.tangentAtLastPoint;
    return this.line(xDir * distance, yDir * distance);
  }

  close(): Diagram {
    const segments = closeSegments(this.pendingSegments);
    return loopySegmentsToDiagram(segments);
  }

  closeWithMirror(): Diagram {
    if (!this.pendingSegments.length) throw new Error("No segments to close");

    const firstSegment = this.pendingSegments[0];
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const lastSegment = this.pendingSegments.at(-1)!;

    const mirrorVector = subtract(
      lastSegment.lastPoint,
      firstSegment.firstPoint
    );
    const mirrorTranform = new TransformationMatrix().mirrorLine(
      mirrorVector,
      firstSegment.firstPoint
    );

    const mirroredSegments = this.pendingSegments.map((segment) =>
      segment.transform(mirrorTranform).reverse()
    );
    mirroredSegments.reverse();

    return loopySegmentsToDiagram([
      ...this.pendingSegments,
      ...mirroredSegments,
    ]);
  }
}

export function draw(origin: Vector = [0, 0]): DrawingPen {
  return new DrawingPen(origin);
}
