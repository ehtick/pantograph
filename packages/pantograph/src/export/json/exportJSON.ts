import { Diagram } from "../../models/Diagram";
import { Figure } from "../../models/Figure";
import { Loop } from "../../models/Loop";
import { Arc } from "../../models/segments/Arc";
import { Line } from "../../models/segments/Line";
import { jsonDiagram } from "./jsonDiagram";
import { jsonFigure } from "./jsonFigure";
import { jsonLoop } from "./jsonLoop";
import { jsonSegment } from "./jsonSegment";

type Shape = Loop | Figure | Diagram | Arc | Line;

export function exportJSON(shape: Shape) {
  if (shape instanceof Diagram) {
    return jsonDiagram(shape);
  } else if (shape instanceof Figure) {
    return jsonFigure(shape);
  } else if (shape instanceof Loop) {
    return jsonLoop(shape);
  } else if (shape instanceof Arc || shape instanceof Line) {
    return jsonSegment(shape);
  } else {
    throw new Error("Unknown shape type");
  }
}
