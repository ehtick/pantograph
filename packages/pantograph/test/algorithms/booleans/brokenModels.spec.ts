import { describe, it, expect } from "vitest";
import { importJSON } from "../../../src/import/json/importJSON";

import { cut } from "../../../src/booleanOperations";

import fixture from "./fixture1.json";

describe("tests that correspond to bugs found in real models", () => {
  it.only("should cut the model in fixture 1", () => {
    const [a, b] = fixture.map(importJSON);
    const cutted = cut(a, b);
    expect(cutted).toMatchSnapshot();
  });
});
