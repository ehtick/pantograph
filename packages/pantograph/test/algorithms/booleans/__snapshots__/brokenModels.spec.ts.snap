// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tests that correspond to bugs found in real models > should cut the model in fixture 1 1`] = `
Diagram {
  "_boundingBox": null,
  "figures": [
    Figure {
      "contour": Loop {
        "_boundingBox": BoundingBox {
          "xMax": 9.30045367892232,
          "xMin": -9.300453678922278,
          "yMax": 38.897163531315805,
          "yMin": 17.416985258030042,
        },
        "_clockwise": null,
        "segments": [
          Arc {
            "_angularLength": 0.06840761595045941,
            "_boundingBox": BoundingBox {
              "xMax": 9.30045367892232,
              "xMin": -9.300453678922278,
              "yMax": 17.57605623427429,
              "yMin": 17.416985258030042,
            },
            "_coefficients": null,
            "_firstAngle": 1.5365925188196663,
            "_lastAngle": 1.6050001347701257,
            "_radius": 271.9658397803519,
            "center": [
              -1.3526572670654706e-13,
              -254.38978354707763,
            ],
            "clockwise": false,
            "firstPoint": [
              9.30045367792232,
              17.41698525903007,
            ],
            "lastPoint": [
              -9.300453677922278,
              17.416985259030042,
            ],
            "precision": 1e-9,
            "segmentType": "ARC",
          },
          Arc {
            "_angularLength": 0.08705411024424858,
            "_boundingBox": BoundingBox {
              "xMax": 1.0000475014143811e-9,
              "xMin": -9.300453678922278,
              "yMax": 38.897163531315805,
              "yMin": 17.416985258030042,
            },
            "_coefficients": null,
            "_firstAngle": 2.7765107218117855,
            "_lastAngle": 2.689456611567537,
            "_radius": 268.96583978035187,
            "center": [
              241.93906132136178,
              -78.61076631141151,
            ],
            "clockwise": true,
            "firstPoint": [
              -9.300453677922278,
              17.416985259030042,
            ],
            "lastPoint": [
              4.750141438099034e-14,
              38.89716353031581,
            ],
            "precision": 1e-9,
            "segmentType": "ARC",
          },
          Arc {
            "_angularLength": 0.08705411024424853,
            "_boundingBox": BoundingBox {
              "xMax": 9.30045367892232,
              "xMin": -9.99952498585619e-10,
              "yMax": 38.897163531315805,
              "yMin": 17.41698525803007,
            },
            "_coefficients": null,
            "_firstAngle": 0.452136042022255,
            "_lastAngle": 0.36508193177800646,
            "_radius": 268.9658397803519,
            "center": [
              -241.93906132136186,
              -78.6107663114112,
            ],
            "clockwise": true,
            "firstPoint": [
              4.750141438099034e-14,
              38.89716353031581,
            ],
            "lastPoint": [
              9.30045367792232,
              17.41698525903007,
            ],
            "precision": 1e-9,
            "segmentType": "ARC",
          },
        ],
        "strokeType": "LOOP",
      },
      "holes": [],
    },
    Figure {
      "contour": Loop {
        "_boundingBox": BoundingBox {
          "xMax": 25.13270200049966,
          "xMin": -25.132702000499915,
          "yMax": 92.5037648683675,
          "yMin": 41.76493318182681,
        },
        "_clockwise": null,
        "segments": [
          Arc {
            "_angularLength": 0.021475252065441985,
            "_boundingBox": BoundingBox {
              "xMax": 7.661146116053629,
              "xMin": 4.8780107618016935,
              "yMax": 53.512212067067644,
              "yMin": 48.45095599931289,
            },
            "_coefficients": null,
            "_firstAngle": 2.649571335313931,
            "_lastAngle": 2.628096083248489,
            "_radius": 268.9658397803519,
            "center": [
              241.93906132136178,
              -78.61076631141151,
            ],
            "clockwise": true,
            "firstPoint": [
              4.8780107628016935,
              48.450956000312885,
            ],
            "lastPoint": [
              7.661146115053629,
              53.51221206606765,
            ],
            "precision": 1e-9,
            "segmentType": "ARC",
          },
          Arc {
            "_angularLength": 0.03084017963648744,
            "_boundingBox": BoundingBox {
              "xMax": 7.661146116053629,
              "xMin": -9.999420907027317e-10,
              "yMax": 53.512212067067644,
              "yMin": 50.33295709076028,
            },
            "_coefficients": null,
            "_firstAngle": 5.121165795606332,
            "_lastAngle": 5.090325615969845,
            "_radius": 268.9658397803519,
            "center": [
              -99.24933830855309,
              300.3173402321163,
            ],
            "clockwise": true,
            "firstPoint": [
              7.661146115053629,
              53.51221206606765,
            ],
            "lastPoint": [
              5.790929726849339e-14,
              50.33295709176028,
            ],
            "precision": 1e-9,
            "segmentType": "ARC",
          },
          Arc {
            "_angularLength": 0.03084017963648833,
            "_boundingBox": BoundingBox {
              "xMax": 1.0000579092972685e-9,
              "xMin": -7.6611461160535725,
              "yMax": 53.512212067067686,
              "yMin": 50.33295709076028,
            },
            "_coefficients": null,
            "_firstAngle": 4.334452344799534,
            "_lastAngle": 4.303612165163045,
            "_radius": 268.9658397803521,
            "center": [
              99.24933830855338,
              300.3173402321164,
            ],
            "clockwise": true,
            "firstPoint": [
              5.790929726849339e-14,
              50.33295709176028,
            ],
            "lastPoint": [
              -7.6611461150535725,
              53.51221206606769,
            ],
            "precision": 1e-9,
            "segmentType": "ARC",
          },
          Arc {
            "_angularLength": 0.021475252065442263,
            "_boundingBox": BoundingBox {
              "xMax": -4.87801076180158,
              "xMin": -7.6611461160535725,
              "yMax": 53.512212067067686,
              "yMin": 48.450955999312875,
            },
            "_coefficients": null,
            "_firstAngle": 0.5134965703413028,
            "_lastAngle": 0.49202131827586054,
            "_radius": 268.9658397803519,
            "center": [
              -241.93906132136186,
              -78.6107663114112,
            ],
            "clockwise": true,
            "firstPoint": [
              -7.6611461150535725,
              53.51221206606769,
            ],
            "lastPoint": [
              -4.87801076280158,
              48.45095600031287,
            ],
            "precision": 1e-9,
            "segmentType": "ARC",
          },
          Arc {
            "_angularLength": 0.07932337102768638,
            "_boundingBox": BoundingBox {
              "xMax": -4.87801076180158,
              "xMin": -25.132702000499915,
              "yMax": 48.45095600131287,
              "yMin": 41.76493318182681,
            },
            "_coefficients": null,
            "_firstAngle": 5.070886143261684,
            "_lastAngle": 4.991562772233998,
            "_radius": 268.9658397803519,
            "center": [
              -99.24933830855309,
              300.3173402321163,
            ],
            "clockwise": true,
            "firstPoint": [
              -4.87801076280158,
              48.45095600031287,
            ],
            "lastPoint": [
              -25.132701999499915,
              41.76493318282681,
            ],
            "precision": 1e-9,
            "segmentType": "ARC",
          },
          Arc {
            "_angularLength": 0.20857412949765397,
            "_boundingBox": BoundingBox {
              "xMax": 9.999524972635898e-10,
              "xMin": -25.132702000499915,
              "yMax": 92.5037648683675,
              "yMin": 41.76493318182681,
            },
            "_coefficients": null,
            "_firstAngle": 5.71898986502995,
            "_lastAngle": 5.927563994527604,
            "_radius": 271.9658397803521,
            "center": [
              -254.94903116637155,
              187.19489177353896,
            ],
            "clockwise": false,
            "firstPoint": [
              -25.132701999499915,
              41.76493318282681,
            ],
            "lastPoint": [
              -4.7502736410236056e-14,
              92.5037648673675,
            ],
            "precision": 1e-9,
            "segmentType": "ARC",
          },
          Arc {
            "_angularLength": 0.20857412949765353,
            "_boundingBox": BoundingBox {
              "xMax": 25.13270200049966,
              "xMin": -1.0000475027364104e-9,
              "yMax": 92.5037648683675,
              "yMin": 41.7649331818269,
            },
            "_coefficients": null,
            "_firstAngle": 3.4972139662417736,
            "_lastAngle": 3.705788095739427,
            "_radius": 271.96583978035204,
            "center": [
              254.94903116637147,
              187.1948917735387,
            ],
            "clockwise": false,
            "firstPoint": [
              -4.7502736410236056e-14,
              92.5037648673675,
            ],
            "lastPoint": [
              25.13270199949966,
              41.764933182826894,
            ],
            "precision": 1e-9,
            "segmentType": "ARC",
          },
          Arc {
            "_angularLength": 0.0793233710276855,
            "_boundingBox": BoundingBox {
              "xMax": 25.13270200049966,
              "xMin": 4.8780107618016935,
              "yMax": 48.45095600131288,
              "yMin": 41.7649331818269,
            },
            "_coefficients": null,
            "_firstAngle": 4.43321518853538,
            "_lastAngle": 4.353891817507694,
            "_radius": 268.9658397803521,
            "center": [
              99.24933830855338,
              300.3173402321164,
            ],
            "clockwise": true,
            "firstPoint": [
              25.13270199949966,
              41.764933182826894,
            ],
            "lastPoint": [
              4.8780107628016935,
              48.450956000312885,
            ],
            "precision": 1e-9,
            "segmentType": "ARC",
          },
        ],
        "strokeType": "LOOP",
      },
      "holes": [],
    },
  ],
}
`;
